rfcs/search_relevancy_sandbox/20230331-project_proposal_search_relevancy_sandbox.md

```diff
@@ -0,0 +1,145 @@
+# Project Proposal: Search Relevancy Sandbox
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] dhruvkb
+- [ ] sarayourfriend
+
+## Project summary
+
+<!-- A brief one or two sentence summary of the project's features -->
+
+Develop mechanisms and workflows for sustainably creating & updating our staging
+API database & Elasticsearch cluster, in order to enable testing changes to
+Elasticsearch indices. The staging setup will reflect production in two ways:
+proportionally by provider and with production-level volumes.
+
+## Goals
+
+<!-- Which yearly goal does this project advance? -->
+
+Yearly goal: **Result Relevancy**
+
+## Requirements
+
+<!-- Detailed descriptions of the features required for the project. Include user stories if you feel they'd be helpful, but focus on describing a specification for how the feature would work with an eye towards edge cases. -->
+
+The following should be available once this project is complete:
+
+1. A mechanism by which maintainers can easily update the staging API database
+   with recent data.
+2. A mechanism by which maintainers can create/update a small but
+   proportional-to-production-by-provider index.
+3. A mechanism by which maintainers can create/update a production-data-volume
+   sized index.
+4. A mechanism by which maintainers can easily deploy & iterate on new
+   Elasticsearch indexing configurations
+5. A mechanism by which maintainers can easily point the staging Elasticsearch
+   index aliases (currently `image` and `audio`) to one of the above.
+
+Many of these mechanisms will likely be built in Airflow.
+
+### Proportional by provider
+
+The proportional-by-provider index will be a subset of the production data,
+constructed so that the number of records by provider is roughly proportional to
+the percentages per provider that exist in production. This index will allow us
+to create tests down the line which might assess result relevancy in an
+automated way, i.e. based on how many results from a particular provider appear
+in results (see the
+[provider tally script](https://github.com/WordPress/openverse/pull/397) for
+examples of these results). This could be done by:
```

Comment:
I'm slightly sceptical that this is a reasonable goal/down-the-line use case. For it to work we'd not only need proportional-by-provider, we'd also need to scope it by specific queries... at which point we're optimising one or a handful of specific query terms. We don't have a uniform distribution, for example, of "dog" images, or "vase", or "flower", etc. I'm sure Flickr is dominant in all of them, but I'd bet few museums have many results for dog but many do for vases and iNaturalist for flowers, etc.

Overall, I don't think we can actually test relevancy in this way. It _needs_ to be based on user data and a/b style testing to see whether click-through increases. Otherwise, we'll be falling back to what I earlier called the "we know it when we see it" approach, which I have low to zero confidence in and is not a quantitative approach viable in the long-term.

It is worth mentioning this to avoid us couching work in the possibility of being able to "manually test relevancy" in staging, either with actual production data or proportionally-similar-to-production data. There are use-cases for manually testing index changes in staging, but I think primarily for testing things like text analysis changes to address specific problems rather than broad relevancy. Those, in particular, would need to have _specific_ data sets, not just proportional-by-provider data sets.

If we wanted to cover changes to the stemmer for the following, how large a percentage of the production data would we need to be able to actually reliably see results relevant to these changes?

- "universe"
- "animal"
- "news"

It's not unreasonable to think that even 50% of production data would exclude one or the other of these, and these are real examples of stemmer issues we have encountered in production. There are likely heaps of others that will be even more diverse.

The bottom line here is that: proportional-by-provider _sounds_ right, it does sound like a good idea. However, I just don't know what it's actually useful for if it's a naive proportion and not directed for specific query terms. At which point, it's only useful for specific problems, and we need to incorporate the assumption that specific query terms are needed to generate these proportional-by-provider indexes.

It might also be useful for something like an integration test that runs through a representative and random sample of data for each provider and tests thumbnails and other edge cases that we've found vary by provider at times. But then we need to decide what "random" means and how we'd implement it in the context of ES and Postgres, I'm sure each have their own unique and variously relevant approaches to that problem.

Reply:
Additional comment though to say that aside from the motivation, the broad approach described in this section makes sense to me based on the assumptions it was written under. I just want to clarify that I am questioning the assumptions for this section, not the implementation overview in it, which for the assumptions given, make sense to me.

Reply:
One more thing: I could see proportional-by-provider (but still statistically relevant sample size!) being useful for iterating on changes to how indexing happens: shard sizes, slicing, etc. In that case, the actual data itself doesn't matter, just that we have a statistically significant sample size that can be used to understand how long, on average, documents take to index, etc. If we want to iterate on the ingestion server configuration as well, we can use this to have shorter but still relevant enough runs.

Forgive my repeated clarifications. I just want to make sure that my line of questioning isn't misunderstood to mean that proportional data is outright not useful. I'm specifically sceptical about its usefulness for relevancy testing which I think, by necessity, needs to be done in production, with real user behaviour.

Reply:
These are great questions and points! I appreciate you drawing question to the assumed need for this type of index at all. My original motivation behind this was actually the [data transfer script](https://github.com/WordPress/openverse-infrastructure/blob/5b3e2c625d6119d4c8a0ceefbfc3bac4e58b82a5/next/modules/concrete/api/scripts/data-transfer.sh) we used when creating a small sample dataset for the API ECS migration. In that case, we selected 100 rows (not randomized) from each provider for exactly the reasons you describe - to check that there were no issues serving them via the API and that the thumbnails worked. 

My _primary_ motivation here is that I want to have an index that is both **small** and **has data from each provider we serve in production**. The former is a requirement in my mind because I don't want to make any dev wait around for one (or several) days while reindexing happens so they can check that something for a provider works in staging. The reasoning for the latter is to ensure that we have many possible edge cases available (though having the entire dataset indexed also nets us this).

I'm not tied to the proportionality piece of this, my intent was to _maybe_ catch two birds with one stone by having this both be small _and_ proportional, but your reasoning for why it may not be useful makes sense.

Do you think there would still be a benefit to having, like we did in API ECS testing, an index that had 100 records per provider? That could replace this index requirement and we could remove the proportionality aspect described here.

Reply:
As I said in the last comment, my main concern is in establishing an expectation that these data sets enable artificial relevancy tests. Relevancy needs to be tested against real user data. My issue is not with the concept of a smaller data set generally. It is undoubtedly useful for a myriad of use cases other than relevancy: measuring indexing performance, integration testing dead link and thumbnail checks with all providers, etc. Proportionality and randomised selection is especially useful for something like index performance as you need a representative sample. Integration tests theoretically require far fewer results but also benefit from sample diversity so that we cast as wide a net as possible, even in a smaller data set.

I fully agree we should have this ability, I just want us to be clear on the motivations and use cases it is relevant to.

Reply:
Ah, I see! Thanks for clarifying your intentions - that makes perfect sense. I'll remove the language around using this to assess result relevancy and replace it with some of the motivations you mention.